File: app/src/main/java/com/example/minesweeper/GameEngine.java
package com.example.minesweeper;

import java.util.Random;

public class GameEngine {
    public static final int EASY_ROWS = 9;
    public static final int EASY_COLS = 9;
    public static final int EASY_MINES = 10;

    public static final int MEDIUM_ROWS = 16;
    public static final int MEDIUM_COLS = 16;
    public static final int MEDIUM_MINES = 40;

    public static final int HARD_ROWS = 16;
    public static final int HARD_COLS = 30;
    public static final int HARD_MINES = 99;

    private Cell[][] grid;
    private int rows;
    private int cols;
    private int totalMines;
    private int revealedCells;
    private int flaggedMines;
    private boolean gameOver;
    private boolean gameWon;
    private boolean firstClick;
    private Random random;

    public interface GameListener {
        void onCellRevealed(int row, int col);
        void onGameOver(boolean won);
        void onMinesCountChanged(int remainingMines);
    }

    private GameListener listener;

    public GameEngine(int difficulty) {
        random = new Random();
        firstClick = true;
        gameOver = false;
        gameWon = false;
        revealedCells = 0;
        flaggedMines = 0;

        switch (difficulty) {
            case 0: // Easy
                rows = EASY_ROWS;
                cols = EASY_COLS;
                totalMines = EASY_MINES;
                break;
            case 1: // Medium
                rows = MEDIUM_ROWS;
                cols = MEDIUM_COLS;
                totalMines = MEDIUM_MINES;
                break;
            case 2: // Hard
                rows = HARD_ROWS;
                cols = HARD_COLS;
                totalMines = HARD_MINES;
                break;
            default:
                rows = EASY_ROWS;
                cols = EASY_COLS;
                totalMines = EASY_MINES;
        }

        grid = new Cell[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                grid[i][j] = new Cell();
            }
        }
    }

    public void setListener(GameListener listener) {
        this.listener = listener;
    }

    public void generateMines(int excludeRow, int excludeCol) {
        int minesPlaced = 0;
        while (minesPlaced < totalMines) {
            int row = random.nextInt(rows);
            int col = random.nextInt(cols);
            
            // 确保第一次点击的位置不是雷
            if (row == excludeRow && col == excludeCol) {
                continue;
            }
            
            if (!grid[row][col].isMine()) {
                grid[row][col].setMine(true);
                minesPlaced++;
            }
        }
        
        // 计算每个格子周围的雷数
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (!grid[i][j].isMine()) {
                    int count = countAdjacentMines(i, j);
                    grid[i][j].setAdjacentMines(count);
                }
            }
        }
    }

    private int countAdjacentMines(int row, int col) {
        int count = 0;
        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                int newRow = row + i;
                int newCol = col + j;
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    if (grid[newRow][newCol].isMine()) {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    public void revealCell(int row, int col) {
        if (gameOver || gameWon) {
            return;
        }

        Cell cell = grid[row][col];
        if (cell.isRevealed() || cell.isFlagged()) {
            return;
        }

        if (firstClick) {
            generateMines(row, col);
            firstClick = false;
        }

        cell.setRevealed(true);
        revealedCells++;

        if (cell.isMine()) {
            gameOver = true;
            if (listener != null) {
                listener.onGameOver(false);
            }
            return;
        }

        if (listener != null) {
            listener.onCellRevealed(row, col);
        }

        // 如果周围没有雷，自动展开
        if (cell.getAdjacentMines() == 0) {
            revealAdjacentCells(row, col);
        }

        // 检查是否获胜
        if (revealedCells == (rows * cols - totalMines)) {
            gameWon = true;
            if (listener != null) {
                listener.onGameOver(true);
            }
        }
    }

    private void revealAdjacentCells(int row, int col) {
        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                int newRow = row + i;
                int newCol = col + j;
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    Cell cell = grid[newRow][newCol];
                    if (!cell.isRevealed() && !cell.isFlagged() && !cell.isMine()) {
                        cell.setRevealed(true);
                        revealedCells++;
                        if (listener != null) {
                            listener.onCellRevealed(newRow, newCol);
                        }
                        if (cell.getAdjacentMines() == 0) {
                            revealAdjacentCells(newRow, newCol);
                        }
                    }
                }
            }
        }
    }

    public void toggleFlag(int row, int col) {
        if (gameOver || gameWon) {
            return;
        }

        Cell cell = grid[row][col];
        if (cell.isRevealed()) {
            return;
        }

        cell.setFlagged(!cell.isFlagged());
        if (cell.isFlagged()) {
            flaggedMines++;
            if (cell.isMine()) {
                // 正确标记的雷
            }
        } else {
            flaggedMines--;
        }

        if (listener != null) {
            listener.onMinesCountChanged(totalMines - flaggedMines);
        }
    }

    public Cell getCell(int row, int col) {
        return grid[row][col];
    }

    public int getRows() {
        return rows;
    }

    public int getCols() {
        return cols;
    }

    public int getTotalMines() {
        return totalMines;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public boolean isGameWon() {
        return gameWon;
    }

    public void revealAllMines() {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j].isMine()) {
                    grid[i][j].setRevealed(true);
                    if (listener != null) {
                        listener.onCellRevealed(i, j);
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/example/minesweeper/GameEngine.kt
package com.example.minesweeper

import kotlin.random.Random

class GameEngine(private val difficulty: Int) {
    companion object {
        const val EASY_ROWS = 9
        const val EASY_COLS = 9
        const val EASY_MINES = 10

        const val MEDIUM_ROWS = 16
        const val MEDIUM_COLS = 16
        const val MEDIUM_MINES = 40

        const val HARD_ROWS = 16
        const val HARD_COLS = 30
        const val HARD_MINES = 99
    }

    interface GameListener {
        fun onCellRevealed(row: Int, col: Int)
        fun onGameOver(won: Boolean)
        fun onMinesCountChanged(remainingMines: Int)
    }

    private val grid: Array<Array<Cell>>
    private val rows: Int
    private val cols: Int
    private val totalMines: Int
    private var revealedCells = 0
    private var flaggedMines = 0
    private var gameOver = false
    private var gameWon = false
    private var firstClick = true
    private val random = Random.Default

    private var listener: GameListener? = null

    init {
        val config = when (difficulty) {
            0 -> Triple(EASY_ROWS, EASY_COLS, EASY_MINES)
            1 -> Triple(MEDIUM_ROWS, MEDIUM_COLS, MEDIUM_MINES)
            2 -> Triple(HARD_ROWS, HARD_COLS, HARD_MINES)
            else -> Triple(EASY_ROWS, EASY_COLS, EASY_MINES)
        }
        rows = config.first
        cols = config.second
        totalMines = config.third

        grid = Array(rows) { Array(cols) { Cell() } }
    }

    fun setListener(listener: GameListener?) {
        this.listener = listener
    }

    fun generateMines(excludeRow: Int, excludeCol: Int) {
        var minesPlaced = 0
        while (minesPlaced < totalMines) {
            val row = random.nextInt(rows)
            val col = random.nextInt(cols)

            // 确保第一次点击的位置不是雷
            if (row == excludeRow && col == excludeCol) {
                continue
            }

            if (!grid[row][col].isMine) {
                grid[row][col].isMine = true
                minesPlaced++
            }
        }

        // 计算每个格子周围的雷数
        for (i in 0 until rows) {
            for (j in 0 until cols) {
                if (!grid[i][j].isMine) {
                    grid[i][j].adjacentMines = countAdjacentMines(i, j)
                }
            }
        }
    }

    private fun countAdjacentMines(row: Int, col: Int): Int {
        var count = 0
        for (i in -1..1) {
            for (j in -1..1) {
                val newRow = row + i
                val newCol = col + j
                if (newRow in 0 until rows && newCol in 0 until cols) {
                    if (grid[newRow][newCol].isMine) {
                        count++
                    }
                }
            }
        }
        return count
    }

    fun revealCell(row: Int, col: Int) {
        if (gameOver || gameWon) return

        val cell = grid[row][col]
        if (cell.isRevealed || cell.isFlagged) return

        if (firstClick) {
            generateMines(row, col)
            firstClick = false
        }

        cell.isRevealed = true
        revealedCells++

        if (cell.isMine) {
            gameOver = true
            listener?.onGameOver(false)
            return
        }

        listener?.onCellRevealed(row, col)

        // 如果周围没有雷，自动展开
        if (cell.adjacentMines == 0) {
            revealAdjacentCells(row, col)
        }

        // 检查是否获胜
        if (revealedCells == (rows * cols - totalMines)) {
            gameWon = true
            listener?.onGameOver(true)
        }
    }

    private fun revealAdjacentCells(row: Int, col: Int) {
        for (i in -1..1) {
            for (j in -1..1) {
                val newRow = row + i
                val newCol = col + j
                if (newRow in 0 until rows && newCol in 0 until cols) {
                    val cell = grid[newRow][newCol]
                    if (!cell.isRevealed && !cell.isFlagged && !cell.isMine) {
                        cell.isRevealed = true
                        revealedCells++
                        listener?.onCellRevealed(newRow, newCol)
                        if (cell.adjacentMines == 0) {
                            revealAdjacentCells(newRow, newCol)
                        }
                    }
                }
            }
        }
    }

    fun toggleFlag(row: Int, col: Int) {
        if (gameOver || gameWon) return

        val cell = grid[row][col]
        if (cell.isRevealed) return

        cell.isFlagged = !cell.isFlagged
        if (cell.isFlagged) {
            flaggedMines++
        } else {
            flaggedMines--
        }

        listener?.onMinesCountChanged(totalMines - flaggedMines)
    }

    fun getCell(row: Int, col: Int): Cell = grid[row][col]

    fun getRows(): Int = rows

    fun getCols(): Int = cols

    fun getTotalMines(): Int = totalMines

    fun isGameOver(): Boolean = gameOver

    fun isGameWon(): Boolean = gameWon

    fun revealAllMines() {
        for (i in 0 until rows) {
            for (j in 0 until cols) {
                if (grid[i][j].isMine) {
                    grid[i][j].isRevealed = true
                    listener?.onCellRevealed(i, j)
                }
            }
        }
    }
}

